corehq/apps/userreports/pillow.py

```diff
@@ -222,6 +225,81 @@ def _save_doc_to_table(self, domain, table, doc, eval_context):
+                )
+
+            doc_store = document_stores[0]
+            docs = list(doc_store.iter_documents([change.id for change in _changes]))
+            docs_by_id = {doc['_id']: doc for doc in docs}
+
+            for change in _changes:
+                if change.id not in docs_by_id:
+                    # we need to capture DocumentMissingError which is not possible in bulk
+                    #   so let pillow fall back to serial mode to capture the error
+                    failed_changes.append(change)
+                try:
+                    ensure_matched_revisions(change, docs_by_id.get(change.id))
+                except DocumentMismatchError:
+                    failed_changes.append(change)
+            good_changes = set(_changes) - failed_changes
+            return [docs_by_id.get(change.id) for change in good_changes]
+
+        self.bootstrap_if_needed()
+
+        changes_by_domain = defaultdict(list)
+        for change in changes:
+            if change.metadata.domain and change.metadata.domain in self.table_adapters_by_domain:
+                # if no domain we won't save to any UCR table
+                changes_by_domain[change.metadata.domain].append(change)
+
+        for domain, changes_chunk in six.iteritems(changes_by_domain):
+            rows_to_save_by_adapter = defaultdict(list)
+            doc_ids_to_delete_by_adapter = defaultdict(set)
+            adapters = list(self.table_adapters_by_domain[domain])
+            to_delete = {change for change in changes_chunk if change.deleted}
+            to_update = set(changes_chunk) - to_delete
+            async_configs_by_doc_id = defaultdict(list)
+            docs = get_docs(to_update, domain)
+
+            for adapter in adapters:
+                doc_ids_to_delete_by_adapter[adapter] = to_delete
+                for doc in docs:
+                    eval_context = EvaluationContext(doc)
+                    if adapter.config.filter(doc):
+                        if adapter.run_asynchronous:
+                            async_configs_by_doc_id[doc['_id']].append(adapter.config._id)
+                        else:
+                            # TOdo; handle error
+                            rows_to_save_by_adapter[adapter].extend(adapter.get_all_values(doc, eval_context))
+                            doc_ids_to_delete_by_adapter[adapter].add(doc['_id'])
+                            eval_context.reset_iteration()
+                    elif adapter.config.deleted_filter(doc) or adapter.doc_exists(doc):
+                        doc_ids_to_delete_by_adapter[adapter].add(doc['_id'])
+
+            for adapter, rows in six.iteritems(rows_to_save_by_adapter):
+                try:
+                    adapter._save_rows(rows, doc_ids_to_delete_by_adapter[adapter])
+                except Exception as e:
+                    raise BulkPorcessingError('Error bulk saving calculated UCR rows {}'.format(e))
+
+            succeeded_changes.union(changes_chunk)
```

Comment:
Some of these could be failures. This should only include changes for docs that were returned by `get_docs`